### PR TITLE
Implement push.

### PR DESCRIPTION
**Pull request by tcard**
This PR adds push notification support to the Ably JS library. Note that Push notifications support is currently in public beta. The API may change before the final release. Please report any bugs or feedback to support@ably.io

_Edit 2019-01-12 (Simon):_

#### Please do not directly checkout this branch any more, it has been superseded by the ably-js 1.1 beta. To use the 1.1 beta:

- For node, or browsers with webpack: install the ably 1.1 beta with `npm install ably@beta`, and require as normal per the readme
- For browsers for use with a `script` tag: source as `<script src="https://cdn.ably.io/lib/ably.min-1.1.0-beta.0.js"></script>`

-----------------------------------

To customers who were previously using the push beta by checking out this branch, please note some API changes between that and the 1.1 beta:
- `deviceRegistrations/channelSubscriptions#list` now does what `#get` used to (take params and return a paginated result)
- `deviceRegistations#get` now takes a `deviceId` or a `DeviceDetails` and returns a `DeviceDetails`
- `channelSubscriptions#get` has been removed
- `deviceRegistrations#removeWhere` does what `#remove` used to do; `#remove` now takes a deviceId or DeviceDetails
- `deviceRegistrations/channelSubscriptions#save` now returns the just-saved object rather than an undecoded buffer
- `RealtimeChannel#push` has been removed; please use the methods in `client.push.admin` instead

### CODE REVIEW

File: browser/lib/transport/xhrrequest.js [hunk 1]
```diff
@@ -101,7 +102,7 @@ var XHRRequest = (function() {
 				xhr.abort();
 			}, timeout),
 			body = this.body,
-			method = body ? 'POST' : 'GET',
+			method = method,
```
Comment (SimonWoolf) [inline browser/lib/transport/xhrrequest.js]: I suspect you meant `method = this.method` here

File: browser/lib/util/http.js [hunk 2]
```diff
@@ -103,17 +109,8 @@ var Http = (function() {
...
-	 * @param the full path of the POST request
-	 * @param headers optional hash of headers
-	 * @param body object or buffer containing request body
-	 * @param params optional hash of params
-	 * @param callback (err, response)
-	 */
-	Http.postUri = function(rest, uri, headers, body, params, callback) {
-		Http.Request(rest, uri, headers, params, body, callback || noop);
+	Http.doUri = function(method, rest, uri, headers, body, params, callback) {
+		Http.Request(method, rest, uri, headers, params, body, callback || noop);
```
Comment (SimonWoolf) [inline browser/lib/util/http.js]: Not a big fan of having both Http.doUri and Http.Request just to swap the order of some parameters. Happy to switch to `do` + `doUri` if you like, so why don't we get rid of `Http.Request` altogether and just make xhrRequest etc. implement `Http.doUri`?

File: common/lib/client/push.js [hunk 3]
```diff
@@ -0,0 +1,134 @@
+var Push = (function() {
+	var msgpack = (typeof require !== 'function') ? Ably.msgpack : require('msgpack-js');
```
Comment (SimonWoolf) [inline common/lib/client/push.js]: `var msgpack = Platform.msgpack`, similarly for other files

File: common/lib/client/push.js [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+	function Push(rest) {
+		this.rest = rest;
+		this.admin = new Admin(rest);
+	}
+
+	Push.prototype.publish = function(recipient, payload, callback) {
+		var rest = this.rest;
+		var format = rest.options.useBinaryProtocol ? 'msgpack' : 'json',
+		    requestBody = Utils.mixin({recipient: recipient}, payload),
+		    headers = Utils.copy(Utils.defaultPostHeaders(format));
```
Comment (SimonWoolf) [inline common/lib/client/push.js]: `Utils.defaultPostHeaders` creates and returns a new object, so I don't think you need the `Utils.copy`

File: common/lib/client/push.js [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+		if(rest.options.headers)
+			Utils.mixin(headers, rest.options.headers);
+
+		requestBody = (format == 'msgpack') ? msgpack.encode(requestBody, true): JSON.stringify(requestBody);
```
Comment (SimonWoolf) [inline common/lib/client/push.js]: Given that's all you're using the `format` variable for, you could get rid of it and just do `requestBody = rest.options.useBinaryProtocol ? msgpack.encode(requestBody, true): JSON.stringify(requestBody)`

File: common/lib/types/pushchannelsubscription.js [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+	};
+
+	PushChannelSubscription.fromResponseBody = function(body, format) {
+		if(format)
+			body = (format == 'msgpack') ? msgpack.decode(body) : JSON.parse(String(body));
+
+		for(var i = 0; i < body.length; i++) {
+			body[i] = PushChannelSubscription.fromDecoded(body[i]);
+		}
+		return body;
```
Comment (SimonWoolf) [inline common/lib/types/pushchannelsubscription.js]: Could this not be replaced with `return PushChannelSubscription.fromValuesArray(body)`, and get rid of `PushChannelSubscription.fromDecoded`?

File: spec/rest/push.test.js [hunk 7]
```diff
@@ -0,0 +1,511 @@
...
+		var x = arguments[1];
+		var y = arguments[2];
+
+		var eq = includesUnordered(x, y);
+		test.ok(eq === true, JSON.stringify(x, null, 2) + ' includesUnordered ' + JSON.stringify(y, null, 2) + ' (' + eq + ')');
+	}
+
+	function isArray(object) {
+ 		return object != null && typeof object === "object" && 'splice' in object && 'join' in object;
+	}
```
Comment (SimonWoolf) [inline spec/rest/push.test.js]: There's a `Utils.isArray`

File: browser/lib/transport/xhrrequest.js (hunk 1 above)
Comment (tcard) [inline browser/lib/transport/xhrrequest.js]: Fixed!

File: browser/lib/util/http.js (hunk 2 above)
Comment (tcard) [inline browser/lib/util/http.js]: Alternatively, I've gotten rid of `Http.doUri`, as it is a smaller change.

File: common/lib/client/push.js (hunk 3 above)
Comment (tcard) [inline common/lib/client/push.js]: Done.

File: common/lib/client/push.js (hunk 4 above)
Comment (tcard) [inline common/lib/client/push.js]: OK, just copypasted this really. Fixed old instances too.

File: common/lib/client/push.js (hunk 5 above)
Comment (tcard) [inline common/lib/client/push.js]: But I'm also using it as an argument to `defaultPostHeaders`.

File: common/lib/types/pushchannelsubscription.js (hunk 6 above)
Comment (tcard) [inline common/lib/types/pushchannelsubscription.js]: Right. Done.

**Comment by SimonWoolf** [comment]
Rebased on top of latest `master`. 

@tcard the push tests fail (not as a result of the rebase), mind fixing whatever's wrong? cheers
```
✖ push_deviceRegistrations_save
Assertion Message: property missing or null: recipient

✖ push_deviceRegistrations_get
Assertion Message: property missing or null: recipient

✖ push_deviceRegistrations_remove
Assertion Message: property missing or null: recipient
```

**Comment by gianpaj** [comment]
Hi,

Does this PR include push notifications on mobile apps using React Native or only web / browser-based notifications?

thanks

**Comment by paddybyers** [comment]
> Does this PR include push notifications on mobile apps using React Native or only web / browser-based notifications?

We've not looked yet at what we would need to do to hook into native (ie GCM/FCM/APNS) push notifications via React native, and this PR does not address that. However, we do delivery push notifications via those transports to native apps, so in theory it shouldn't be too hard to enable via React.

**Comment by SimonWoolf** [comment]
I've rebased this on latest master and merged into `integration-1-1`. Leaving this PR open so people can still use the beta until will do a full release.

**Comment by paddybyers** [comment]
> I've rebased this on latest master and merged into `integration-1-1`. Leaving this PR open so people can still use the beta until will do a full release.

@SimonWoolf I think this branch has been superseded. There is a separate branch (https://github.com/ably/ably-js/tree/push-activation) with various other fixes to the push implementation and tests.

**Comment by paddybyers** [comment]
> I've rebased this on latest master and merged into `integration-1-1`. Leaving this PR open so people can still use the beta until will do a full release.

@SimonWoolf I think this branch has been superseded. There is a separate branch (https://github.com/ably/ably-js/tree/push-activation) with various other fixes to the push implementation and tests.

**Comment by paddybyers** [comment]
> I've rebased this on latest master and merged into `integration-1-1`. Leaving this PR open so people can still use the beta until will do a full release.

@SimonWoolf I think this branch has been superseded. There is a separate branch (https://github.com/ably/ably-js/tree/push-activation) with various other fixes to the push implementation and tests.

**Comment by paddybyers** [comment]
> I've rebased this on latest master and merged into `integration-1-1`. Leaving this PR open so people can still use the beta until will do a full release.

@SimonWoolf I think this branch has been superseded. There is a separate branch (https://github.com/ably/ably-js/tree/push-activation) with various other fixes to the push implementation and tests.

**Comment by paddybyers** [comment]
> I've rebased this on latest master and merged into `integration-1-1`. Leaving this PR open so people can still use the beta until will do a full release.

@SimonWoolf I think this branch has been superseded. There is a separate branch (https://github.com/ably/ably-js/tree/push-activation) with various other fixes to the push implementation and tests.

**Comment by paddybyers** [comment]
> I've rebased this on latest master and merged into `integration-1-1`. Leaving this PR open so people can still use the beta until will do a full release.

@SimonWoolf I think this branch has been superseded. There is a separate branch (https://github.com/ably/ably-js/tree/push-activation) with various other fixes to the push implementation and tests.

**Comment by paddybyers** [comment]
> I've rebased this on latest master and merged into `integration-1-1`. Leaving this PR open so people can still use the beta until will do a full release.

@SimonWoolf I think this branch has been superseded. There is a separate branch (https://github.com/ably/ably-js/tree/push-activation) with various other fixes to the push implementation and tests.

**Comment by SimonWoolf** [comment]
> I think this branch has been superseded. There is a separate branch (https://github.com/ably/ably-js/tree/push-activation) with various other fixes to the push implementation and tests.

That's a descendant of where this branch was prior to the rebase, so we can take the useful bits from that branch and cherry-pick into `integration-1-1` without conflict when we're ready

**Pull request closed without merging**